### PR TITLE
Build LCNC-EMC on MK-HAL

### PR DESCRIPTION
**Pull request by zultron**
Following #279, #280 and #281, this is the fourth of four PRs to support building LCNC-EMC on MK-HAL.

This PR prepares for a [matching fork][1] of LCNC to build its EMC against this tree.  This PR reverts some minor API changes made in Machinekit since the fork from LCNC, and updates the build system and packaging for that purpose.  It creates a `machinekit-hal.pc` file with all needed build parameters needed on the LCNC side to configure its build against an external HAL by specifying `./configure --with-hal=machinekit-hal`.

This branch is very near completion, but still needs help in the top two commits to fully integrate with the new CI system.  For now, they're hacked to demonstrate the full build:  MK-HAL package build and install; LCNC Docker image build; LCNC build against MK

[1]: https://github.com/zultron/machinekit/tree/zultron/2019-07-03-2.8-mk-hal-build

### CODE REVIEW

**Comment by zultron** [comment]
OK, so CI is passing, but it revives the `build-preview.yaml` and hobbles the `debian-builder-workflow.yaml` workflows.  I'm hoping @cerna can advise here.

**Comment by cerna** [comment]
@zultron,
personally I don't think that the flow of logic (of build, to be precise) should be done this way. Don't get me wrong, it is good as a prototype (that's why I said that in the e-mail), but in production, I don't think that Machinekit-HAL should be tested against the LinuxCNC, but LinuxCNC (being a module to Machinekit-HAL) should be tested against Machinekit-HAL.

I presume here that the packages of EMC3 (my provisional name for the LinuxCNC module) will have hard dependency on specific version of Machinekit-HAL package.

So from my point of view, the course of action should be this: create a new repository in Machinekit organization (aptly named something like `EMC3` or `LinuxCNCModule` or so [I would like opinions here]) with `master` branch with cron job every day rebasing it to LinuxCNC@master, then with branch `module` where everything needed to run atop Machinekit-HAL would be. (This `module` branch would be rebased atop `master`).

When building the EMC3, similar action like @cmcquinn is doing in branch [package-ci](https://github.com/cmcquinn/machinekit-cnc/tree/package-ci) - build a new Docker image with Machinekit-HAL-Debian-Builder and then build the packages. So repository EMC3 would have another 11 GitHub Packages with names like `emc3-debian-builder-v.${tag}:latest`.

The problem with this is pull requests: You would need to assume that user doesn't break building of the LinuxCNC module with his pull request to Machinekit-HAL@master and if so that he opens a new pull request against a EMC3@module.

So, in conclusion, given that this pull request only turn off `debian-builder-workflow.yaml` and turns back on the (very changed) temporary one (i.e. nothing which would be visible forever like new GitHub packages), I would merge it as is, then set up the other repository and I then will implement what I described.

Opinions?

**Comment by zultron** [comment]
That sounds reasonable to me, and I don't even want the temporary workflow to be merged, so I've removed that top commit.

**Comment by zultron** [comment]
The Travis CI build failed because the Docker image needs updating.

**Comment by cerna** [comment]
Yeah, I am surprised that the Travis integration graced this pull request with its presence. Normally it just doesn't work and plays dead. Did you do something to it?

I will regenerate images as soon as possible. (Or do something else about it.)

I have seen that you were playing with Github Actions cache and Artifact upload for Docker images. Not sure what conclusions you reached, but mine were that it is unusable for use with Docker images of sizes that Machinekit-HAL uses. Cache has maximum size of 5 GB before it starts overwriting itself and maximum storage life of one week. And artifact upload never worked reliably for me. Plus I had to `XZ` the exported archive with the insane settings, which took 30 minutes to just archive.

BTW, what name of the LinuxCNC clone repository do **you** want?

**Comment by cerna** [comment]
The co-installation problem was solved:
```
sudo apt install machinekit-hal machinekit-hal-rt-preempt machinekit-hal-dbgsym machinekit-hal-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Suggested packages:
  hostmot2-firmware-all
The following NEW packages will be installed:
  machinekit-hal-dev
The following packages will be upgraded:
  machinekit-hal machinekit-hal-dbgsym machinekit-hal-rt-preempt
3 upgraded, 1 newly installed, 0 to remove and 57 not upgraded.
3 not fully installed or removed.
Need to get 12.4 MB of archives.
After this operation, 7,567 kB disk space will be freed.
Get:1 https://dl.cloudsmith.io/public/machinekit/machinekit-hal/deb/debian buster/main amd64 machinekit-hal-dbgsym amd64 0.3.1589042880.git56a42a9-1~buster [10.7 MB]
Get:2 https://dl.cloudsmith.io/public/machinekit/machinekit-hal/deb/debian buster/main amd64 machinekit-hal amd64 0.3.1589042880.git56a42a9-1~buster [1,488 kB]
Get:3 https://dl.cloudsmith.io/public/machinekit/machinekit-hal/deb/debian buster/main amd64 machinekit-hal-rt-preempt amd64 0.3.1589042880.git56a42a9-1~buster [2,942 B]
Get:4 https://dl.cloudsmith.io/public/machinekit/machinekit-hal/deb/debian buster/main amd64 machinekit-hal-dev amd64 0.3.1589042880.git56a42a9-1~buster [244 kB]
Fetched 12.4 MB in 6s (1,987 kB/s)                                                                                                                                                                 
Reading changelogs... Done
(Reading database ... 981360 files and directories currently installed.)
Preparing to unpack .../machinekit-hal-dbgsym_0.3.1589042880.git56a42a9-1~buster_amd64.deb ...
Unpacking machinekit-hal-dbgsym (0.3.1589042880.git56a42a9-1~buster) over (0.3.1588961152.git10f5a51-1~buster) ...
Preparing to unpack .../machinekit-hal_0.3.1589042880.git56a42a9-1~buster_amd64.deb ...
Unpacking machinekit-hal (0.3.1589042880.git56a42a9-1~buster) over (0.3.1588961152.git10f5a51-1~buster) ...
Preparing to unpack .../machinekit-hal-rt-preempt_0.3.1589042880.git56a42a9-1~buster_amd64.deb ...
Unpacking machinekit-hal-rt-preempt (0.3.1589042880.git56a42a9-1~buster) over (0.3.1588961152.git10f5a51-1~buster) ...
Preparing to unpack .../machinekit-hal-dev_0.3.1589042880.git56a42a9-1~buster_amd64.deb ...
Unpacking machinekit-hal-dev (0.3.1589042880.git56a42a9-1~buster) ...
Setting up machinekit-hal (0.3.1589042880.git56a42a9-1~buster) ...
Valid MKUUID in mkuuid 39b7d24d-a7a3-486e-947e-be1db9a3b8a7 - using that
Old value in /etc/machinekit/machinekit.ini replaced with new MKUUID 39b7d24d-a7a3-486e-947e-be1db9a3b8a7
Setting up machinekit-hal-dbgsym (0.3.1589042880.git56a42a9-1~buster) ...
Setting up machinekit-hal-dev (0.3.1589042880.git56a42a9-1~buster) ...
Setting up machinekit-hal-rt-preempt (0.3.1589042880.git56a42a9-1~buster) ...
Processing triggers for rsyslog (8.1901.0-1) ...
Processing triggers for libc-bin (2.28-10) ...
```
Thank you.

**Comment by zultron** [comment]
> Yeah, I am surprised that the Travis integration graced this pull request with its presence. Normally it just doesn't work and plays dead. Did you do something to it?

I didn't.  IIRC, someone in the "Owner" group should be able to go over to Travis-CI.org and shut off builds for this repo.

> I have seen that you were playing with Github Actions cache and Artifact upload for Docker images. Not sure what conclusions you reached, but mine were that it is unusable for use with Docker images of sizes that Machinekit-HAL uses. Cache has maximum size of 5 GB before it starts overwriting itself and maximum storage life of one week. And artifact upload never worked reliably for me. Plus I had to `XZ` the exported archive with the insane settings, which took 30 minutes to just archive.

Failed experiment.  I thought I remembered you could `docker load` an image and it would serve as cache for a `docker build`, but I was wrong.

> BTW, what name of the LinuxCNC clone repository do **you** want?

I don't have a firm opinion on this, but let's take the discussion over to #267.
